src/transformers/models/relation_detr/modeling_relation_detr.py

```diff
@@ -0,0 +1,2871 @@
+        """Input shape: Batch x Time x Channel"""
+
+        batch_size, target_len, embed_dim = hidden_states.size()
+        # add position embeddings to the hidden states before projecting to queries and keys
+        hidden_states_original = hidden_states
+        if position_embeddings is not None:
+            hidden_states = self.with_pos_embed(hidden_states, position_embeddings)
+
+        # get queries, keys and values
+        query_states = self.q_proj(hidden_states) * self.scaling
+        key_states = self._shape(self.k_proj(hidden_states), -1, batch_size)
+        value_states = self._shape(self.v_proj(hidden_states_original), -1, batch_size)
+
+        proj_shape = (batch_size * self.num_heads, -1, self.head_dim)
+        query_states = self._shape(query_states, target_len, batch_size).view(*proj_shape)
+        key_states = key_states.view(*proj_shape)
+        value_states = value_states.view(*proj_shape)
+
+        source_len = key_states.size(1)
+
+        attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
+
+        if attn_weights.size() != (batch_size * self.num_heads, target_len, source_len):
+            raise ValueError(
+                f"Attention weights should be of size {(batch_size * self.num_heads, target_len, source_len)}, but is"
+                f" {attn_weights.size()}"
+            )
+
+        # expand attention_mask
+        if attention_mask is not None:
+            if attention_mask.dtype == torch.bool:
+                float_attention_mask = torch.zeros_like(attention_mask, dtype=hidden_states.dtype)
+                float_attention_mask.masked_fill_(~attention_mask, float("-inf"))
+                attention_mask = float_attention_mask
+
+            if attention_mask.dim() == 2:
+                if attention_mask.size() != (target_len, source_len):
+                    raise ValueError(
+                        f"2D-Attention mask should be of size {(target_len, source_len)}, but is {attention_mask.size()}"
+                    )
+                # [seq_len, seq_len] -> [batch_size, self.num_heads, target_seq_len, source_seq_len]
+                attention_mask = attention_mask.expand(batch_size, self.num_heads, *attention_mask.size())
+            elif attention_mask.dim() == 3:
+                if attention_mask.size() != (batch_size * self.num_heads, target_len, source_len):
+                    raise ValueError(
+                        f"3D-Attention mask should be of size {(batch_size, self.num_heads, target_len, source_len)}, but"
+                        f" is {attention_mask.size()}"
+                    )
+                attention_mask = attention_mask.view(batch_size, self.num_heads, target_len, source_len)
+            else:
+                raise ValueError(
+                    f"Attention mask should be of size {(target_len, source_len)} of {(batch_size * self.num_heads, target_len, source_len)}, but is"
+                    f" {attention_mask.size()}"
+                )
```

Comment:
I think we can not prepare it in advance. Because `attention_mask` has different values and dims in different branches and layers of Relation-DETR. This is where relation-detr differs from other models, it adds position relation to `attention_mask`